File: selenium_mcp/tools/network.py
"""Network monitoring tools."""

import logging
from pydantic import BaseModel, Field

from ..tool_base import BaseTool, ToolSchema, ToolResult
from ..context import Context

logger = logging.getLogger(__name__)

class NetworkParams(BaseModel):
    """Parameters for network operations."""
    action: str = Field(description="Network action: 'get_requests', 'clear', 'set_offline'")
    offline: bool = Field(default=False, description="Set offline mode (for set_offline action)")

class NetworkTool(BaseTool):
    """Monitor and control network activity."""
    
    def _create_schema(self) -> ToolSchema:
        return ToolSchema(
            name="network_monitor",
            description="Monitor network requests or control network state",
            input_schema=NetworkParams,
            tool_type="destructive"
        )
    
    async def handle(self, context: Context, params: NetworkParams) -> ToolResult:
        """Handle network operations."""
        driver = context.current_tab_or_die()
        
        async def network_action():
            if params.action == "get_requests":
                try:
                    # Get performance logs if available
                    logs = driver.get_log('performance')
                    network_requests = []
                    
                    for log in logs:
                        message = log.get('message', {})
                        if isinstance(message, str):
                            import json
                            try:
                                message = json.loads(message)
                            except:
                                continue
                                
                        method = message.get('message', {}).get('method', '')
                        if 'Network.' in method:
                            params_data = message.get('message', {}).get('params', {})
                            if 'request' in params_data:
                                url = params_data['request'].get('url', '')
                                method_type = params_data['request'].get('method', '')
                                network_requests.append(f"{method_type} {url}")
                    
                    logger.info(f"🌐 Retrieved {len(network_requests)} network requests")
                    return "\n".join(network_requests) if network_requests else "No network requests found"
                    
                except Exception as e:
                    logger.warning(f"Could not retrieve network requests: {e}")
                    return "Network monitoring not available"
                    
            elif params.action == "clear":
                try:
                    # Clear performance logs
                    driver.get_log('performance')
                    logger.info("🌐 Cleared network logs")
                    return "Network logs cleared"
                except Exception as e:
                    logger.warning(f"Could not clear network logs: {e}")
                    return "Network logs clear failed"
                    
            elif params.action == "set_offline":
                try:
                    # Set network offline mode using Chrome DevTools
                    driver.execute_cdp_cmd('Network.enable', {})
                    driver.execute_cdp_cmd('Network.emulateNetworkConditions', {
                        'offline': params.offline,
                        'latency': 0,
                        'downloadThroughput': -1,
                        'uploadThroughput': -1
                    })
                    
                    mode = "offline" if params.offline else "online"
                    logger.info(f"🌐 Set network mode to {mode}")
                    return f"Network set to {mode} mode"
                except Exception as e:
                    logger.warning(f"Could not set network mode: {e}")
                    return "Network mode change failed"
            else:
                raise ValueError(f"Invalid network action: {params.action}")
        
        # Robot Framework code
        if params.action == "get_requests":
            code = [
                "# Get network requests",
                "# Note: Network request monitoring not directly supported in Robot Framework"
            ]
        elif params.action == "clear":
            code = [
                "# Clear network logs", 
                "# Note: Network log clearing not directly supported in Robot Framework"
            ]
        elif params.action == "set_offline":
            mode = "offline" if params.offline else "online"
            code = [
                f"# Set network to {mode} mode",
                "# Note: Network mode control not directly supported in Robot Framework"
            ]
        else:
            code = [
                f"# Network action: {params.action}",
                "# Note: Network operations limited in Robot Framework"
            ]
        
        return ToolResult(
            code=code,
            action=network_action,
            capture_snapshot=False,
            wait_for_network=False
        )

File: selenium-mcp-server/src/tools/browser/network.ts
import { z } from 'zod';
import { BaseTool } from '../base.js';
import { Context } from '../../context.js';
import { ToolResult } from '../../types.js';

const schema = z.object({
  action: z.enum(['get_requests', 'clear', 'set_offline']).describe('Network action'),
  offline: z.boolean().optional().default(false).describe('Set offline mode (for set_offline action)')
});

export class NetworkTool extends BaseTool {
  readonly name = 'network_monitor';
  readonly description = 'Monitor network requests or control network state';
  readonly inputSchema = schema;

  async execute(context: Context, params: unknown): Promise<ToolResult> {
    const { action, offline } = this.parseParams(schema, params);

    const driver = await context.getDriver();

    if (action === 'get_requests') {
      try {
        const logs = await driver.manage().logs().get('performance');
        const networkRequests: string[] = [];

        for (const log of logs) {
          try {
            const message = JSON.parse(log.message);
            const method = message?.message?.method || '';
            if (method.includes('Network.')) {
              const params = message?.message?.params;
              if (params?.request) {
                const reqMethod = params.request.method || '';
                const url = params.request.url || '';
                networkRequests.push(`${reqMethod} ${url}`);
              }
            }
          } catch {
            // Skip unparseable logs
          }
        }

        if (networkRequests.length === 0) {
          return this.success('No network requests found', false);
        }

        return this.success(`Network requests:\n${networkRequests.join('\n')}`, false);
      } catch (err) {
        return this.success('Network monitoring not available', false);
      }
    } else if (action === 'clear') {
      try {
        await driver.manage().logs().get('performance');
        return this.success('Network logs cleared', false);
      } catch (err) {
        return this.error('Network logs clear failed');
      }
    } else if (action === 'set_offline') {
      try {
        // Use Chrome DevTools Protocol to set network conditions
        await (driver as any).sendDevToolsCommand('Network.enable', {});
        await (driver as any).sendDevToolsCommand('Network.emulateNetworkConditions', {
          offline: offline,
          latency: 0,
          downloadThroughput: -1,
          uploadThroughput: -1
        });

        const mode = offline ? 'offline' : 'online';
        return this.success(`Network set to ${mode} mode`, false);
      } catch (err) {
        return this.error('Network mode change failed (requires Chrome)');
      }
    }

    return this.error(`Invalid network action: ${action}`);
  }
}
